cruft.8
.TH CRUFT 8
.SH NAME
cruft \- Check the filesystem for cruft (missing and unexplained files)
.SH SYNOPSIS
.B cruft [ FILE ]
.SH "DESCRIPTION"
.B cruft
is a program that goes over the Debian packaging system's database, and
compares the contents of that with the files actually on your system, and
produces a summary of the differences.
.SH OPTIONS
If a file is provided, cruft\-ng will analyze this single file
and print matching package(s) name(s).
.br
Without any option, cruft\-ng will analyze the whole system.
.SH "FILES"
A user can tune execution by providing extra configuration files:
.TP
.B /etc/cruft/ignore
Do not process paths listed in this file, this can speed-up matching.
Format is one line per path.
Example values: /var/schroot/, /var/www/ ...
.TP
.B  /etc/cruft/filters/*
Rules (glob-patterns) similar that are similar to those built-in into
.I cruft
package.
The uppercase files are always processed, the lowercase ones are processed
when matching package is installed.
.TP
.B /etc/cruft/explain/*
Shell scripts that will print-out a list of temporary file
belonging to a package.
The uppercase files are always processed, the lowercase ones are processed
when matching package is installed.

.SH "ENVIRONMENT VARIABLES"
It is possible to get some debugging information from cruft by setting the variable
.B DEBUG.
.SH "CAVEAT"
The cleansing command
.B apt purge '~c'
should be run beforehand. Beware that this may cleanse old databases
and currently unused settings you would have liked to preserve.
.SH "SEE ALSO"
Documentation from original cruft package:
https://github.com/a-detiste/cruft/blob/master/README
.SH AUTHOR
This manual page was written by Alexandre Detiste <alexandre@detiste.be>,
for the Debian GNU/Linux system (but may be used by others).


debian/changelog
cruft-ng (0.9.72) unstable; urgency=medium

  * Integration the results of https://openqa.debian.net/ testing.
  * Add rules for gnome-brave-icon-theme, gnome-colors-common,
    gnome-dust-icon-theme, gnome-human-icon-theme,
    gnome-illustrious-icon-theme, gnome-noble-icon-theme,
    gnome-wine-icon-theme, gnome-wise-icon-theme, slick-greeter,
    mate-icon-theme
  * Update rules for: colord, gnome-accessibility-themes, lightdm
  * Refresh GPL-2 license text

  [ Christian Göttsche ]
  * update podman

 -- Alexandre Detiste <tchet@debian.org>  Mon, 21 Jul 2025 16:29:31 +0200

cruft-ng (0.9.71) unstable; urgency=medium

  * Add rules for libc6-x32, libc6-i386.
  * Update rules for emacsen-common, petsc, sddm, grub-efi-amd64-bin.
  * Remove or archive old rules: intel-microcode, debtags,
    apt-xapian-index, wine-development, libapache2-mod-php8.3.

  [ Jochen Sprickerhof ]
  * Use apt for purge "~c" example in the man page
  * Drop /var/log/{lastlog,btmp,wtmp} from base-files and logrotate
  * Add rules for firebird4.0-common and lcl-utils-4.0

 -- Alexandre Detiste <tchet@debian.org>  Fri, 13 Jun 2025 23:04:18 +0200

cruft-ng (0.9.70) unstable; urgency=medium

  * Test Trixie RC1 with default Gnome desktop
  * Add rule for: gnome-browser-connector
  * Update rule for installation-report, init-system-helpers,
    packagekit, update gdm3

  [ Christian Göttsche ]
  * rspamd: include fonts copied into www document root
  * php-common: add module root directory

 -- Alexandre Detiste <tchet@debian.org>  Sun, 18 May 2025 18:54:05 +0200

cruft-ng (0.9.69) unstable; urgency=medium

  * Add rules for rtpengine-daemon, dodgindiamond2,
    python3.14 (from experimental)
  * Update nfs-common
  * Remove rule for Python 3.12, never was in a stable release
  * wtmpdb: the library will create the DB by itsell, rename to libwtmpdb0

  [ NoisyCoil ]
  * add local-apt-repository.sources to local-apt-repository
    (Closes: #1104253)

 -- Alexandre Detiste <tchet@debian.org>  Sat, 17 May 2025 16:35:08 +0200

cruft-ng (0.9.68) unstable; urgency=medium

  * Add rules for ghc-doc, gnubg, glhack, casacore-data*, fp-compiler-3.2.2
  * Update rules for libuuid1

  [ Christian Göttsche ]
  * Add rules for podman and valkey (from redis ones) (Closes: #1102166)
  * Update rules for postfix, php-fpm, certbot, rspamd
  * Fix typos

 -- Alexandre Detiste <tchet@debian.org>  Sun, 13 Apr 2025 12:49:08 +0200

cruft-ng (0.9.67) unstable; urgency=medium

  * Add rules for mallard-rng, elpa-notmuch
  * Update rules for git, plasma-browser-integration
  * Update Standards-Version to 4.7.0

  [ Jochen Sprickerhof ]
  * Add /var/lib/dpkg/arch-native for dpkg/1.22.16
  * Update raspi-firmware

 -- Alexandre Detiste <tchet@debian.org>  Sat, 29 Mar 2025 20:10:13 +0100

cruft-ng (0.9.66) unstable; urgency=medium

  * Add new rules: animals, breeze-cursor-theme, dbeaver-ce,
    gnustep-base-common, freesweep, kf6-breeze-icon-theme,
    libapache2-mod-php8.4, libpaper2, odoo-18, varnish, vmm,
    webmin, wing, xbomb, xletters.
  * Update rules: apt, clamav-freshclam, dovecote-core, sddm.
  * Archive old rules: breeze-icon-theme, libpaper1, llvm-15-runtime,
    jupyter-nbextension-jupyter-js-widgets, qemu-user-static
  * Reformat d/control with debputy

 -- Alexandre Detiste <tchet@debian.org>  Fri, 07 Mar 2025 12:05:31 +0100

cruft-ng (0.9.65) unstable; urgency=medium

  * Package is now team maintained by Debian Security Tools team.
    It can be used as a forensic tool to investigate systems
    messed-up by previous owner/admins.
  * Add new rules: pcp, greed, gtkballs
  * Update rules: init-system-helpers, ltris, grafana, installation-report

 -- Alexandre Detiste <tchet@debian.org>  Sun, 10 Nov 2024 15:18:50 +0100

cruft-ng (0.9.64) unstable; urgency=medium

  * Fix bug found by Ubuntu CI (Closes: #1081002) (LP: #2079840)
  * new rules: tango-common, yorick, docbook-xsl-ns, docbook5-xml,
    python3.13, kxd, libopenmpi-dev, mecab-ipadic, mecab-ipadic-utf8,
    memcached, glx-alternative-nvidia, nvidia-kernel-dkms, octave
  * updated rules: LINUX-IMAGE, ocaml, sddm, wtpmdb
  * archive rule for python3.11
  * update Standards-Version to 4.7.0

  [ Jochen Sprickerhof ]
  * nolocate: don't inspect /run
  * debug: list unused ruleset entries
  * Move /var/log/lastlog to base-files
  * Update ocaml-base
  * Add rules for u-boot-menu and journal-brief

 -- Alexandre Detiste <tchet@debian.org>  Fri, 13 Sep 2024 18:55:45 +0200

cruft-ng (0.9.63) unstable; urgency=medium

  * welcome Jochen Sprickerhof as co-maintainer
  * add new rules: gunicorn, gravitywars, ketm, spacearyarya,
    live-boot, live-config, swtpm-tools, apparmor, udhcpc
  * update existing rules: base-files, python3, raspi-firmware,
    sddm, systemd, ofono, libvirt-daemon-config-network,
    clonezilla, lightdm, system-tools-backends, gnome-session

  [ Jochen Sprickerhof ]
  * new "--no-locate" option to make cruft-ng work without locate
    (Closes: #1072389)
  * new "--root" option to inspect root file systems from outside. Explain
    scripts need to check $CRUFT_ROOT to check inside or skip. Only
    implemented for "--no-locate".
  * add new rules: mpd-sima, fonts-urw-base35, lmodern, alsa-ucm-conf

 -- Alexandre Detiste <tchet@debian.org>  Mon, 17 Jun 2024 16:09:49 +0200

cruft-ng (0.9.62) unstable; urgency=medium

  * run plocate with "--null" option to avoid being confused
    by paths containing the newline character (Closes: #1070216)
  * add new rules: click, clonezilla, drbl,
    libvirt-daemon-config-nwfilter, libvirt-daemon-system,
    monsterz, plasma-browser-integration
  * update rules: anacron, virtualbox-dkms, zfs-dkms
  * 64-bit time transition:
    * libglib2.0      -> libglib2.0-t64
    * libgtk2.0-0     -> libgtk2.0-0t64

 -- Alexandre Detiste <tchet@debian.org>  Sat, 04 May 2024 22:07:25 +0200

cruft-ng (0.9.61) unstable; urgency=medium

  * add new rules: extrepo, syslog-ng-core, w3c-sgml-lib

cruft.cc
// Copyright © 2015 Alexandre Detiste <alexandre@detiste.be>
// SPDX-License-Identifier: GPL-2.0-or-later

#include <iostream>
#include <fstream>
#include <algorithm>
#include <ctime>
#include <thread>

#include <sys/stat.h>
#include <getopt.h>
#include <cstring>
#include <unistd.h>
#include <dirent.h>

#include "explain.h"
#include "filters.h"
#include "locate.h"
#include "dpkg.h"
#include "dpkg_exclude.h"
#include "shellexp.h"
#include "bugs.h"

using namespace std;

#ifdef BUSTER
#define LOCATE_DB "/var/lib/mlocate/mlocate.db"
#define UPDATEDB "updatedb.mlocate"
#else
#include "nolocate.h"
#define LOCATE_DB "/var/lib/plocate/plocate.db"
#define UPDATEDB "updatedb.plocate"
#endif

/*
static int read_mounts(const vector<string>& prunefs, vector<string>& mounts)
{
	// this doesn't include "/", as it always exists
	ifstream mtab("/proc/mounts");
	string mount,type;
	while (mtab.good())
	{
		getline(mtab,mount,' '); // discard device
		getline(mtab,mount,' ');
		getline(mtab,type,' ');

		bool match=false;
		if (mount=="/") match=true;
		else if (mount.rfind("/sys/", 0) == 0) match=true;
		else if (mount.rfind("/dev/", 0) == 0) match=true;
		else {
			for (const auto& elem: prunefs) {
				string uppercase=type;
				transform(uppercase.begin(), uppercase.end(), uppercase.begin(), ::toupper);
				if (uppercase==elem) {
			  		match=true;
					break;
		      	}
	      	}
		}

		if (!match) {
			//cerr << mount << " => " << type << endl;
			mounts.emplace_back(mount);
		}
		getline(mtab,mount); // discard options
	}
	return 0;
}
*/

static bool updatedb()
{
	/* return value is meant as an are_we_up_to_date flag
	   when running as non-root */
	const string db = LOCATE_DB;

	int rc_locate, rc_dpkg;
	struct stat stat_locate, stat_dpkg;
	rc_locate = stat(db.c_str(), &stat_locate);
	rc_dpkg = stat("/var/lib/dpkg/status", &stat_dpkg);

	if (rc_dpkg) {
		cerr << "can't read /var/lib/dpkg/status timestamp !!!\n";
		exit(1);
	}

	if (!rc_locate && stat_locate.st_mtim.tv_sec > stat_dpkg.st_mtim.tv_sec)
		return true;

	if (getuid()) return false;

	if (system(UPDATEDB)) {
		cerr << "updatedb failed\n";
		exit(1);
	}
	return true;
}

static void one_file(const string& path)
{
	struct stat buffer;
	if (stat(path.c_str(), &buffer)) {
		cerr << "file not found\n";
		exit(1);
	}

	// is it a static file ?
	dpkg_start("/");
	if (query(path.c_str())) {
		dpkg_end();
		exit(0);
	}

	string infile = path;
	if ((path.rfind("/bin/", 0) == 0)
         or (path.rfind("/lib/", 0) == 0)
         or (path.rfind("/sbin/", 0) == 0)) {
		infile = "/usr" + infile;
	}

	char* file = realpath(infile.c_str(), nullptr);

	vector<string> packages;
	read_dpkg_header(packages);
	dpkg_end();

	// is it a dynamic file ?
	vector<owner> globs;
	read_filters("/etc/cruft/filters/", "/usr/share/cruft/ruleset", packages, globs);
	for (const auto& gl: globs) {
		if (myglob(file, gl.path)) {
			cout << gl.package << '\n';
			exit(0);
		};
	}

	// match the dynamic "explain" filters
	vector<owner> explain;
	read_explain("/etc/cruft/explain/", packages, explain);
	for (const auto& ex: explain) {
		if (infile == ex.path) {
			cout << ex.package << '\n';
			exit(0);
                }
        }

	cerr << "no matching package found\n";
	exit(1);
}

static int one_package(const string& package)
{
	vector<string> packages;
	packages.push_back(package);
	vector<owner> cruft;
	read_filters("/etc/cruft/filters/", "/usr/share/cruft/ruleset", packages, cruft);
	packages.push_back("binfmt-support");
	packages.push_back("ucf");
	read_explain("/etc/cruft/explain/", packages, cruft);
	for (const auto& cr: cruft) {
		if (cr.package == package) cout << cr.path << '\n';
	}
	return 0;
}

static clock_t beg = clock();

static void elapsed(const string& action)
{
	if (getenv("ELAPSED") == nullptr) return;
	clock_t end = clock();
	clock_t elapsed_mseconds = (end - beg) * 1000 / CLOCKS_PER_SEC;
	cerr << "elapsed " << action << ": " << elapsed_mseconds << '\n';
	beg = end;
}

static const char* const default_explain_dir = "/etc/cruft/explain/";
static const char* const default_filter_dir = "/etc/cruft/filters/";
static const char* const default_ignore_file = "/etc/cruft/ignore";
static const char* const default_ruleset_file = "/usr/share/cruft/ruleset";
static const char* const default_bugs_file = "/usr/share/cruft/bugs";
static const char* const default_root_dir = "/";

static void print_help_message()
{
	cout << "cruft-ng [OPTIONS] [file]\n\n";
	cout << "if <file> is specified, this file is analysed\n";
	cout << "if not, the whole system is analysed\n\n";

	cout << "OPTIONS\n";
	cout << "    -p --package     list volatile files only for this one package\n";
	cout << "    -E --explain     directory for explain scripts (default: " << default_explain_dir << ")\n";
	cout << "    -F --filter      directory for filters (default: " << default_filter_dir << ")\n";
	cout << "    -I --ignore      path for ignore file (default: " << default_ignore_file << ")\n";
	cout << "    -R --ruleset     path for ruleset file (default: " << default_ruleset_file << ")\n";
	cout << "    -B --bugs        path for known bugs file (default: " << default_bugs_file << ")\n";
#ifndef BUSTER
	cout << "    -n --no-locate   do not use locate\n";
	cout << "    -r --root        root directory (default: " << default_root_dir << ", only works with --no-locate)\n";
#endif

	cout << '\n';

	cout << "    -h --help        this help message\n";
}

static void cruft(const string& ignore_file,
                  const string& filter_dir,
                  const string& ruleset_file,
                  const string& explain_dir,
                  const string& bugs_file,
                  bool locate,
                  const string& root_dir)
{
	bool debug = getenv("DEBUG") != nullptr;

	const int SIZEBUF = 200;
	char buf[SIZEBUF];
	time_t rawtime;
	struct tm * timeinfo;
	time(&rawtime);
	timeinfo=localtime(&rawtime);
	setlocale(LC_TIME, "");
	strftime(buf, sizeof(buf), "%c", timeinfo);
	cout << "cruft report: " << buf << '\n';

	if (locate) {
		bool updated = updatedb();
		if (!updated) {
			cerr << "warning: plocate database is outdated" << endl << flush;
		}
		elapsed("updatedb");
	}

	vector<string> fs;
#ifndef BUSTER
	thread thr_plocate(locate ? read_locate : read_nolocate, ref(fs), ignore_file, root_dir);
#else
	thread thr_plocate(read_locate, ref(fs), ignore_file, root_dir);
#endif

	vector<string> packages;
	vector<string> dpkg;
	dpkg_start(root_dir);
	thread thr_dpkg(read_dpkg, ref(packages), ref(dpkg), false, root_dir);
	thr_plocate.join();
	thr_dpkg.join();
	dpkg_end();
	elapsed("plocate + dpkg");

	map<string, bug> bugs;
	read_bugs(bugs, bugs_file);

	// match two main data sources
	vector<string> cruft;
	vector<string> missing;
	for (auto left=fs.begin(), right=dpkg.begin(); left != fs.end() && right != dpkg.end();)
	{
		//cerr << "[" << *left << "=" << *right << "]" << endl;
		if (*left==*right) {
			left++;
			right++;
		} else if (*left < *right) {
			cruft.push_back(*left);
			left++;
		} else {
			missing.push_back(*right);
			right++;
		}
		if (right == dpkg.end()) while(left  !=fs.end()  ) {cruft.push_back(*left);    left++; }
		if (left  == fs.end()  ) while(right !=dpkg.end()) {missing.push_back(*right); right++;}
	}
	elapsed("main set match");

	if (debug) cerr << missing.size() << " files in missing database\n";
	if (debug) cerr << cruft.size() << " files in cruft database\n\n";

	// https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=619086
	vector<string> excludes;
	read_dpkg_excludes(excludes);
	elapsed("read excludes");
	vector<string> missing2;
	unsigned long count_stat = 0;
	for (const auto& miss: missing) {
		bool match=false;
		for (const auto& ex: excludes) {
			match=myglob(miss,ex);
			if (match) break;
		}
		if (!match) {
			// file may exist on tmpfs
			// e.g.: /var/cache/apt/archives/partial
			struct stat stat_buffer;
	                if ( stat(miss.c_str(), &stat_buffer) == 0) {
				count_stat += 1;
				if (debug) cerr << miss << " was not in plocate database\n";
			} else {
				missing2.push_back(miss);
			}
		}
	}
	elapsed("missing2");
	if (debug) cerr << "count stat():" << count_stat << '\n';

	// match the globs against reduced database
	vector<owner> globs;
	read_filters(filter_dir, ruleset_file, packages, globs);
	elapsed("read filters");
	vector<string> cruft3;
	vector<bool> used_globs(globs.size(), false);
	for (const auto& cr: cruft) {
		bool match=false;
		for (size_t i = 0; i < globs.size(); i++) {
			match=myglob(cr, globs[i].path);
			if (match) {
				used_globs[i] = true;
				break;
			}
		}
		if (!match) cruft3.push_back(cr);
	}
	if (debug)
		for (size_t i = 0; i < globs.size(); i++)
			if (!used_globs[i])
				cout << "unused ruleset: " << globs[i].package << " " << globs[i].path << endl;
	elapsed("extra vs globs");
	if (debug) cerr << cruft3.size() << " files in cruft3 database\n\n";

	// set CRUFT_ROOT for explain scripts
	setenv("CRUFT_ROOT", root_dir == "/" ? "" : root_dir.c_str(), 0);
	// match the dynamic "explain" filters
	vector<owner> explain;
	read_explain(explain_dir, packages, explain);
	elapsed("read explain");
	vector<string> cruft4;
	for (const auto& cr: cruft3) {
		bool match=false;
		for (const auto& ex: explain) {
			match=(cr==ex.path);
			if (match) break;
		}
		if (!match) cruft4.push_back(cr);
	}
	elapsed("extra vs explain");

	if (debug) cerr << cruft4.size() << " files in cruft4 database\n";

	//TODO: some smarter algo when run as non-root
        //      like checking the R/X bits of parent dir
	cout << "---- missing: dpkg ----\n";
	if (geteuid() == 0 ) for (const auto& miss: missing2) {
		cout << "        " << miss << '\n';
	}

	//TODO: split by filesystem
	cout << "---- unexplained: / ----\n";
	for (const auto& cr: cruft4) {
		cout << "        " << cr;
		auto bug = bugs.find(cr);
		if (bug != bugs.end()) {
			cout << "       (Bug: #" << bug->second.bugno << ")";
		}
		cout << '\n';
	}

	cout << "\nend.\n";
	exit(0);
}

int main(int argc, char *argv[])
{
	bool do_one_package = false;
	bool locate = true;
	string package = "";
	string explain_dir = default_explain_dir;
	string filter_dir = default_filter_dir;
	string ignore_file = default_ignore_file;
	string ruleset_file = default_ruleset_file;
	string bugs_file = default_bugs_file;
	string root_dir = default_root_dir;

	const struct option long_options[] =
	{
		{"help", no_argument, nullptr, 'h'},
		{"no-locate", no_argument, nullptr, 'n'},
		{"package", required_argument, nullptr, 'p'},
		{"explain", required_argument, nullptr, 'E'},
		{"filter", required_argument, nullptr, 'F'},
		{"ignore", required_argument, nullptr, 'I'},
		{"ruleset", required_argument, nullptr, 'R'},
		{"bugs", required_argument, nullptr, 'B'},
		{"root", required_argument, nullptr, 'r'},
		{0, 0, 0, 0}
	};

	int opt, opti = 0;
	while ((opt = getopt_long(argc, argv, "p:E:F:hI:nR:B:r:", long_options, &opti)) != 0) {
		if (opt == EOF)
			break;

		switch (opt) {
		case 'p':
			do_one_package = true;
			package = optarg;
			break;
		case 'E':
			explain_dir = optarg;
			if (!explain_dir.empty() && explain_dir.back() != '/')
				explain_dir += '/';
			break;

		case 'F':
			filter_dir = optarg;
			if (!filter_dir.empty() && filter_dir.back() != '/')
				filter_dir += '/';
			break;

		case 'h':
			print_help_message();
			exit(0);

		case 'I':
			ignore_file = optarg;
			break;

		case 'n':
			locate = false;
			break;

		case 'R':
			ruleset_file = optarg;
			break;

		case 'B':
			bugs_file = optarg;
			break;

		case '?':
			print_help_message();
			exit(1);

		case 'r':
			root_dir = optarg;
			if (!root_dir.empty() && root_dir.back() != '/')
				root_dir += '/';
			break;

	        default:
	            cerr << "Invalid getopt return value: " << opt << "\n";
				break;
		}
	}


	if (do_one_package) exit(one_package(package));

	if (optind < argc) {
		if (optind + 1 == argc)
			one_file(argv[1]);

		cerr << "Invalid non-option arguments:";
		while (optind < argc)
			cerr << " " << argv[optind++];
		cerr << '\n';
		print_help_message();
		exit(1);
	}

	// else: standard cruft report
	cruft(ignore_file, filter_dir, ruleset_file, explain_dir, bugs_file, locate, root_dir);
}
